enhancements/manifestlist/manifestlist-support.md

```diff
@@ -0,0 +1,277 @@
+So that I can keep my storage resource usage under control
+
+### Implementation Details/Notes/Constraints
+
+#### On pushing manifests into integrated registry
+
+Integrated registry will need to accept image pushes by sha, not only by tag.
+The difference here is that the registry is not supposed to create an Image
+object every time it sees a push by sha but only when it sees a push by tag.
+On the latter scenario the registry may need to create more than one image
+(if the pushed manifest is of type manifest list).
+
+### Risks and Mitigations
+
+#### Console usage of ImageStreamImport
+
+OpenShift console uses ImageStreamImport to obtain information about images
+in order to present them to users. When dealing with manifest lists not all
+information will be presented back to the console (the Config portion of an
+image is absent in a manifest list). Depending on how manifest lists support
+is going to be implemented in the console a second request to the API may be
+needed to select which, among "child images", the user wants to see detailed
+information.
+
+#### Image registry pull through authorization
+
+The current Image and ImageStream layouts are leveraged by the image registry
+when verifying if one is allowed or not to execute a pull through for a certain
+blob. APIs needed to be adjusted to take into account this extra level of
+indirection.
+
+Before providing the client with content image registry goes through a series
+of checks to verify if a valid reference for the blob (by sha) exists and is
+valid within the namespace (repository). Image registry currently caches this
+information but some changes are going to be needed to validate using the newly
+introduced "child images".
+
+## Design Details
+
+### Proposed Image API changes
+
+Images are indexed (named) in OpenShift by the hash of their content. Currently
+an Image is represented by the following structure:
+
+```go
+type Image struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	DockerImageReference string
+	DockerImageMetadata DockerImage
+	DockerImageMetadataVersion string
+	DockerImageManifest string
+	DockerImageLayers []ImageLayer
+	Signatures []ImageSignature
+	DockerImageSignatures [][]byte
+	DockerImageManifestMediaType string
+	DockerImageConfig string
+}
+```
+
+This document proposes the introduction of an extra field pointing to all
+"child images":
+
+```go
+	DockerImageManifests []ImageManifest
+```
+
+`ImageManifest` struct would look like the following:
+
+```go
+type ImageManifest struct {
+	Digest       string
+	MediaType    string
+	ManifestSize int64
+	Architecture string
+	OS           string
+	Variant      string
+}
+```
+
+The Image pointed by the `Digest` property contain all regular information
+(i.e. Config) we can see nowadays in a regular Image.
+
+Not all fields in an Image are relevant for a manifest list, therefore other
+proposed change is to add an "omitempty" flag on `DockerImageLayers` property
```

Comment:
I don't think this is a backward compatible change. Old clients (oc v4.(y-1).z) may expect them to be always present.

Reply:
What is the main source of concern here?

1. `DockerImageLayers` being an empty slice or
2. `DockerImageLayers` not even being presented